File: java/gemini/p2_nc.java
public class p2_nc {

    public static boolean copyMessage(String source, byte[] dest) {
        if (source == null || dest == null) {
            return false; // Failure: Null input
        }

        byte[] sourceBytes = source.getBytes();
        int len = sourceBytes.length;

        if (len > dest.length) {
            return false; // Failure: Destination too small
        }

        System.arraycopy(sourceBytes, 0, dest, 0, len);
        return true; // Success
    }

    public static void main(String[] args) {
        byte[] dest = new byte[20];
        String msg1 = "Hello World";
        String msg2 = "This message is way too long for the buffer";

        // Valid call
        if (copyMessage(msg1, dest)) {
            System.out.println("Success: Copied message.");
        } else {
            System.out.println("Failure: Copy failed.");
        }

        // Invalid call (too long)
        if (copyMessage(msg2, dest)) {
            System.out.println("Success: Copied message.");
        } else {
            System.out.println("Failure: Copy rejected (too long).");
        }
    }
}

File: python/gemini/p2_nc.py
def copy_message(source: str, dest: bytearray) -> bool:
    if source is None or dest is None:
        return False # Failure: Null input
    
    try:
        source_bytes = source.encode('utf-8')
        length = len(source_bytes)
        
        if length > len(dest):
            return False # Failure: Destination too small
            
        dest[:length] = source_bytes
        return True # Success
    except Exception:
        return False

if __name__ == "__main__":
    dest = bytearray(20)
    msg1 = "Hello World"
    msg2 = "This message is way too long for the buffer"

    # Valid call
    if copy_message(msg1, dest):
        print("Success: Copied message.")
    else:
        print("Failure: Copy failed.")

    # Invalid call (too long)
    if copy_message(msg2, dest):
        print("Success: Copied message.")
    else:
        print("Failure: Copy rejected (too long).")
